lib/ansible/module_utils/urls.py

```diff
@@ -831,175 +386,111 @@ def missing_gzip_error():
+        # Handle disabled redirects
+        elif follow_redirects in ('no', 'none', False):
+            raise urllib.error.HTTPError(newurl, code, msg, headers, fp)
 
-    msg.append('You can use validate_certs=False if you do'
-               ' not need to confirm the servers identity but this is'
-               ' unsafe and not recommended.'
-               ' Paths checked for this platform: %s.')
+        method = req.get_method()
 
-    if exc:
-        msg.append('The exception msg was: %s.' % to_native(exc))
+        # Handle non-redirect HTTP status or invalid follow_redirects
+        if follow_redirects in ('all', 'yes', True):
+            if code < 300 or code >= 400:
+                raise urllib.error.HTTPError(req.get_full_url(), code, msg, headers, fp)
+        elif follow_redirects == 'safe':
+            if code < 300 or code >= 400 or method not in ('GET', 'HEAD'):
+                raise urllib.error.HTTPError(req.get_full_url(), code, msg, headers, fp)
+        else:
+            raise urllib.error.HTTPError(req.get_full_url(), code, msg, headers, fp)
 
-    raise SSLValidationError(' '.join(msg) % (hostname, port, ", ".join(paths)))
+        data = req.data
+        origin_req_host = req.origin_req_host
 
+        # Be conciliant with URIs containing a space
+        newurl = newurl.replace(' ', '%20')
 
-def atexit_remove_file(filename):
-    if os.path.exists(filename):
-        try:
-            os.unlink(filename)
-        except Exception:
-            # just ignore if we cannot delete, things should be ok
-            pass
+        # Support redirect with payload and original headers
+        if code in (307, 308):
+            # Preserve payload and headers
+            req_headers = req.headers
+        else:
+            # Do not preserve payload and filter headers
+            data = None
+            req_headers = {k: v for k, v in req.headers.items()
+                           if k.lower() not in ("content-length", "content-type", "transfer-encoding")}
+
+            # http://tools.ietf.org/html/rfc7231#section-6.4.4
+            if code == 303 and method != 'HEAD':
+                method = 'GET'
+
+            # Do what the browsers do, despite standards...
+            # First, turn 302s into GETs.
+            if code == 302 and method != 'HEAD':
+                method = 'GET'
+
+            # Second, if a POST is responded to with a 301, turn it into a GET.
+            if code == 301 and method == 'POST':
+                method = 'GET'
+
+        return urllib.request.Request(
+            newurl,
+            data=data,
+            headers=req_headers,
+            origin_req_host=origin_req_host,
+            unverifiable=True,
+            method=method.upper(),
+        )
 
 
-def make_context(cafile=None, cadata=None, ciphers=None, validate_certs=True, client_cert=None, client_key=None):
+def make_context(cafile=None, cadata=None, capath=None, ciphers=None, validate_certs=True, client_cert=None,
```

Comment:
Are we ok with this potentially breaking calls with positional args? It's not a `_` method but theoretically others could be using this outside of the util.

Reply:
I did check all collections from galaxy, and none were using it. But I'll double check, and if necessary reorder things.